File: csharp/examples/contacts_manager/contacts_manager.cs
using Uimd;

public sealed record Contact(
    string Id,
    string FirstName,
    string LastName,
    string Email,
    string Phone,
    string Company,
    string City,
    string Category,
    bool Active,
    string Notes);

public sealed class ContactForm : ContactFormUI
{
    private string contactId = "";

    public Action<Contact?>? OnClosed { get; set; }

    public ContactForm(Contact? contact = null)
    {
        if (contact is not null)
        {
            contactId = contact.Id;
            first_name.SetValue(contact.FirstName);
            last_name.SetValue(contact.LastName);
            email.SetValue(contact.Email);
            phone.SetValue(contact.Phone);
            company.SetValue(contact.Company);
            city.SetValue(contact.City);
            SelectCategory(contact.Category);
            active.SetChecked(contact.Active);
            notes.SetValue(contact.Notes);
        }
    }

    protected override void onSaveBtnClick()
    {
        if (string.IsNullOrWhiteSpace(first_name.Value) ||
            string.IsNullOrWhiteSpace(last_name.Value) ||
            string.IsNullOrWhiteSpace(email.Value))
        {
            message.SetText("Required fields are missing.");
            return;
        }
        OnClosed?.Invoke(new Contact(
            contactId,
            first_name.Value,
            last_name.Value,
            email.Value,
            phone.Value,
            company.Value,
            city.Value,
            category.SelectedText,
            active.Checked,
            notes.Value));
        CloseWindow();
    }

    protected override void onCancelBtnClick()
    {
        OnClosed?.Invoke(null);
        CloseWindow();
    }

    private void SelectCategory(string value)
    {
        int index = category.Options.IndexOf(value);
        category.SetSelectedIndex(index >= 0 ? index : 0);
    }
}

public sealed class ContactsManager : ContactsManagerUI
{
    private readonly List<Contact> contactRecords = new();
    private bool dirty;
    private bool finished;
    private string deleteContactId = "";

    public ContactsManager()
    {
        Reload();
    }

    protected override void onContactsSelectionChange(List<string> value)
    {
        RefreshDetail();
    }

    protected override void onAddBtnClick()
    {
        OpenForm(null);
    }

    protected override void onEditBtnClick()
    {
        Contact? contact = SelectedContact();
        if (contact is null)
        {
            SetStatus("Select a contact to edit.");
            return;
        }
        OpenForm(contact);
    }

    protected override void onDeleteBtnClick()
    {
        Contact? contact = SelectedContact();
        if (contact is null)
        {
            SetStatus("Select a contact to delete.");
            return;
        }
        deleteContactId = contact.Id;
        MessageBoxYesNo dialog = new("Delete Contact", "Delete " + RowLabel(contact) + "?");
        dialog.OnClose = DeleteConfirmed;
        OpenWindow(dialog);
    }

    protected override void onSaveBtnClick()
    {
        Save();
    }

    protected override void onReloadBtnClick()
    {
        Reload();
    }

    protected override void onCloseBtnClick()
    {
        finished = true;
    }

    protected override bool shouldClose()
    {
        return finished;
    }

    private void Reload()
    {
        contactRecords.Clear();
        contactRecords.AddRange(LoadContacts());
        dirty = false;
        RefreshList(SelectedId());
        SetStatus("Loaded " + contactRecords.Count + " contacts.");
    }

    private void Save()
    {
        SaveContacts(contactRecords);
        dirty = false;
        SetStatus("Saved " + contactRecords.Count + " contacts.");
    }

    private void OpenForm(Contact? contact)
    {
        ContactForm form = new(contact);
        form.OnClosed = OnFormClosed;
        OpenWindow(form);
    }

    private void OnFormClosed(Contact? result)
    {
        if (result is null)
        {
            SetStatus("Edit cancelled.");
            return;
        }
        if (string.IsNullOrEmpty(result.Id))
        {
            string id = NextId().ToString();
            Contact contact = result with { Id = id };
            contactRecords.Add(contact);
            SetStatus("Contact added.");
            dirty = true;
            RefreshList(id);
            return;
        }
        for (int index = 0; index < contactRecords.Count; ++index)
        {
            if (contactRecords[index].Id == result.Id)
            {
                contactRecords[index] = result;
                SetStatus("Contact updated.");
                dirty = true;
                RefreshList(result.Id);
                return;
            }
        }
        dirty = true;
        RefreshList(result.Id);
    }

    private void DeleteConfirmed(bool yes)
    {
        if (!yes)
        {
            SetStatus("Delete cancelled.");
            return;
        }
        int before = contactRecords.Count;
        contactRecords.RemoveAll(contact => contact.Id == deleteContactId);
        deleteContactId = "";
        if (contactRecords.Count == before)
        {
            SetStatus("Contact no longer exists.");
            return;
        }
        dirty = true;
        RefreshList("");
        SetStatus("Contact deleted.");
    }

    private void RefreshList(string selectedContactId)
    {
        List<string> labels = contactRecords.Select(RowLabel).ToList();
        header.SetText("Contacts Manager - " + labels.Count + " records");
        SetContactsList(labels, selectedContactId);
        RefreshDetail();
    }

    private void SetContactsList(List<string> labels, string selectedContactId)
    {
        contacts.SetOptions(labels);
        if (labels.Count == 0)
        {
            contacts.SetSelectedIndex(-1);
            return;
        }
        int selectedIndex = 0;
        if (!string.IsNullOrEmpty(selectedContactId))
        {
            int index = contactRecords.FindIndex(contact => contact.Id == selectedContactId);
            selectedIndex = index >= 0 ? index : 0;
        }
        contacts.SetSelectedIndex(selectedIndex);
    }

    private void RefreshDetail()
    {
        Contact? contact = SelectedContact();
        detail.SetText(contact is null ? "No contact selected." : DetailText(contact));
    }

    private Contact? SelectedContact()
    {
        if (contactRecords.Count == 0)
        {
            return null;
        }
        int index = Math.Clamp(contacts.SelectedIndex, 0, contactRecords.Count - 1);
        return contactRecords[index];
    }

    private string SelectedId()
    {
        return SelectedContact()?.Id ?? "";
    }

    private int NextId()
    {
        int result = 1;
        foreach (Contact contact in contactRecords)
        {
            if (int.TryParse(contact.Id, out int id))
            {
                result = Math.Max(result, id + 1);
            }
        }
        return result;
    }

    private void SetStatus(string text)
    {
        status.SetText(text + (dirty ? " *" : ""));
    }

    private static string RowLabel(Contact contact)
    {
        return (contact.FirstName + " " + contact.LastName).Trim();
    }

    private static string DetailText(Contact contact)
    {
        return contact.FirstName + " " + contact.LastName +
            "\nEmail: " + contact.Email +
            "\nPhone: " + contact.Phone +
            "\nCompany: " + contact.Company +
            "\nCity: " + contact.City +
            "\nCategory: " + contact.Category + " (" + (contact.Active ? "active" : "inactive") + ")" +
            "\n\n" + contact.Notes;
    }

    private static string DataPath()
    {
        string? path = Environment.GetEnvironmentVariable("UI_CONTACTS_MANAGER_DATA_PATH");
        return string.IsNullOrEmpty(path) ? Path.Combine(AppContext.BaseDirectory, "data", "contacts.csv") : path;
    }

    private static List<Contact> LoadContacts()
    {
        List<Contact> result = new();
        string path = DataPath();
        if (!File.Exists(path))
        {
            return result;
        }
        foreach (string line in File.ReadLines(path).Skip(1))
        {
            string[] cells = line.Split(',');
            if (cells.Length < 10)
            {
                continue;
            }
            result.Add(new Contact(cells[0], cells[1], cells[2], cells[3], cells[4], cells[5], cells[6], cells[7], cells[8] is "1" or "true", cells[9]));
        }
        return result;
    }

    private static void SaveContacts(List<Contact> contacts)
    {
        string path = DataPath();
        Directory.CreateDirectory(Path.GetDirectoryName(path) ?? ".");
        List<string> lines = new()
        {
            "id,first_name,last_name,email,phone,company,city,category,active,notes",
        };
        lines.AddRange(contacts.Select(contact =>
            string.Join(",", contact.Id, contact.FirstName, contact.LastName, contact.Email, contact.Phone,
                contact.Company, contact.City, contact.Category, contact.Active ? "true" : "false", contact.Notes)));
        File.WriteAllLines(path, lines);
    }
}

public static class Program
{
    public static int Main(string[] args)
    {
        return GeneratedWindowRuntime.RunGeneratedAppMain(() =>
        {
            ContactsManager app = new();
            return GeneratedWindowRuntime.RunGeneratedWindow(app, app.RuntimeOptions(), args);
        });
    }
}

File: java/examples/contacts_manager/ContactsManager.java
import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.function.Consumer;

import uimd.GeneratedWindowRuntime;
import uimd.MessageBoxYesNo;

public final class ContactsManager extends ContactsManagerUI
{
    private static final String DATA_PATH_ENV = "UI_CONTACTS_MANAGER_DATA_PATH";
    private static final String CSV_HEADER =
        "id,first_name,last_name,email,phone,company,city,category,active,notes";

    private final List<Contact> contactRecords = new ArrayList<>();
    private boolean dirty;
    private boolean finished;
    private String deleteContactId = "";

    public ContactsManager()
    {
        reload();
    }

    @Override
    protected void onContactsSelectionChange(List<String> value)
    {
        refreshDetail();
    }

    @Override
    protected void onAddBtnClick()
    {
        openForm(null);
    }

    @Override
    protected void onEditBtnClick()
    {
        Contact contact = selectedContact();
        if (contact == null)
        {
            setStatus("Select a contact to edit.");
            return;
        }
        openForm(contact);
    }

    @Override
    protected void onDeleteBtnClick()
    {
        Contact contact = selectedContact();
        if (contact == null)
        {
            setStatus("Select a contact to delete.");
            return;
        }
        deleteContactId = contact.id();
        openWindow(new MessageBoxYesNo(
            "Delete Contact",
            "Delete " + rowLabel(contact) + "?",
            result -> deleteConfirmed("yes".equals(result))));
    }

    @Override
    protected void onSaveBtnClick()
    {
        save();
    }

    @Override
    protected void onReloadBtnClick()
    {
        reload();
    }

    @Override
    protected void onCloseBtnClick()
    {
        finished = true;
    }

    @Override
    protected boolean shouldClose()
    {
        return finished;
    }

    private void reload()
    {
        contactRecords.clear();
        contactRecords.addAll(loadContacts());
        dirty = false;
        refreshList(selectedId());
        setStatus("Loaded " + contactRecords.size() + " contacts.");
    }

    private void save()
    {
        saveContacts(contactRecords);
        dirty = false;
        setStatus("Saved " + contactRecords.size() + " contacts.");
    }

    private void openForm(Contact contact)
    {
        ContactForm form = new ContactForm(contact);
        form.setOnClosed(this::onFormClosed);
        openWindow(form);
    }

    private void onFormClosed(Contact result)
    {
        if (result == null)
        {
            setStatus("Edit cancelled.");
            return;
        }
        if (result.id().isEmpty())
        {
            String id = Integer.toString(nextId());
            contactRecords.add(result.withId(id));
            setStatus("Contact added.");
            dirty = true;
            refreshList(id);
            return;
        }
        for (int index = 0; index < contactRecords.size(); ++index)
        {
            if (contactRecords.get(index).id().equals(result.id()))
            {
                contactRecords.set(index, result);
                setStatus("Contact updated.");
                dirty = true;
                refreshList(result.id());
                return;
            }
        }
        dirty = true;
        refreshList(result.id());
    }

    private void deleteConfirmed(boolean yes)
    {
        if (!yes)
        {
            setStatus("Delete cancelled.");
            return;
        }
        int before = contactRecords.size();
        contactRecords.removeIf(contact -> contact.id().equals(deleteContactId));
        deleteContactId = "";
        if (contactRecords.size() == before)
        {
            setStatus("Contact no longer exists.");
            return;
        }
        dirty = true;
        refreshList("");
        setStatus("Contact deleted.");
    }

    private void refreshList(String selectedContactId)
    {
        List<String> labels = contactRecords.stream().map(ContactsManager::rowLabel).toList();
        header.setText("Contacts Manager - " + labels.size() + " records");
        setContactsList(labels, selectedContactId);
        refreshDetail();
    }

    private void setContactsList(List<String> labels, String selectedContactId)
    {
        contacts.setOptions(labels);
        if (labels.isEmpty())
        {
            contacts.setSelectedIndex(-1);
            return;
        }
        int selectedIndex = 0;
        if (selectedContactId != null && !selectedContactId.isEmpty())
        {
            for (int index = 0; index < contactRecords.size(); ++index)
            {
                if (contactRecords.get(index).id().equals(selectedContactId))
                {
                    selectedIndex = index;
                    break;
                }
            }
        }
        contacts.setSelectedIndex(selectedIndex);
    }

    private void refreshDetail()
    {
        Contact contact = selectedContact();
        detail.setText(contact == null ? "No contact selected." : detailText(contact));
    }

    private Contact selectedContact()
    {
        if (contactRecords.isEmpty())
        {
            return null;
        }
        int index = Math.max(0, Math.min(contacts.selectedIndex(), contactRecords.size() - 1));
        return contactRecords.get(index);
    }

    private String selectedId()
    {
        Contact contact = selectedContact();
        return contact == null ? "" : contact.id();
    }

    private int nextId()
    {
        int result = 1;
        for (Contact contact : contactRecords)
        {
            try
            {
                result = Math.max(result, Integer.parseInt(contact.id()) + 1);
            }
            catch (NumberFormatException ignored)
            {
                // Non-numeric identifiers do not participate in numeric allocation.
            }
        }
        return result;
    }

    private void setStatus(String text)
    {
        status.setText(text + (dirty ? " *" : ""));
    }

    private static String rowLabel(Contact contact)
    {
        return (contact.firstName() + " " + contact.lastName()).trim();
    }

    private static String detailText(Contact contact)
    {
        return contact.firstName() + " " + contact.lastName()
            + "\nEmail: " + contact.email()
            + "\nPhone: " + contact.phone()
            + "\nCompany: " + contact.company()
            + "\nCity: " + contact.city()
            + "\nCategory: " + contact.category()
            + " (" + (contact.active() ? "active" : "inactive") + ")"
            + "\n\n" + contact.notes();
    }

    private static Path dataPath()
    {
        String configured = System.getenv(DATA_PATH_ENV);
        if (configured != null && !configured.isEmpty())
        {
            return Path.of(configured);
        }
        return Path.of("python", "examples", "contacts_manager", "data", "contacts.csv");
    }

    private static List<Contact> loadContacts()
    {
        Path path = dataPath();
        if (!Files.exists(path))
        {
            return List.of();
        }
        try
        {
            List<Contact> result = new ArrayList<>();
            List<String> lines = Files.readAllLines(path, StandardCharsets.UTF_8);
            for (int index = 1; index < lines.size(); ++index)
            {
                List<String> cells = splitCsvLine(lines.get(index));
                if (cells.size() < 10)
                {
                    continue;
                }
                result.add(new Contact(
                    cells.get(0),
                    cells.get(1),
                    cells.get(2),
                    cells.get(3),
                    cells.get(4),
                    cells.get(5),
                    cells.get(6),
                    cells.get(7),
                    toBoolean(cells.get(8)),
                    cells.get(9)));
            }
            return result;
        }
        catch (IOException error)
        {
            throw new UncheckedIOException(error);
        }
    }

    private static void saveContacts(List<Contact> records)
    {
        Path path = dataPath();
        try
        {
            Path parent = path.getParent();
            if (parent != null)
            {
                Files.createDirectories(parent);
            }
            List<String> lines = new ArrayList<>();
            lines.add(CSV_HEADER);
            for (Contact contact : records)
            {
                lines.add(String.join(",",
                    csvCell(contact.id()),
                    csvCell(contact.firstName()),
                    csvCell(contact.lastName()),
                    csvCell(contact.email()),
                    csvCell(contact.phone()),
                    csvCell(contact.company()),
                    csvCell(contact.city()),
                    csvCell(contact.category()),
                    contact.active() ? "true" : "false",
                    csvCell(contact.notes())));
            }
            Files.write(path, lines, StandardCharsets.UTF_8);
        }
        catch (IOException error)
        {
            throw new UncheckedIOException(error);
        }
    }

    private static List<String> splitCsvLine(String line)
    {
        List<String> cells = new ArrayList<>();
        StringBuilder cell = new StringBuilder();
        boolean quoted = false;
        for (int index = 0; index < line.length(); ++index)
        {
            char value = line.charAt(index);
            if (value == '"')
            {
                if (quoted && index + 1 < line.length() && line.charAt(index + 1) == '"')
                {
                    cell.append(value);
                    ++index;
                }
                else
                {
                    quoted = !quoted;
                }
            }
            else if (value == ',' && !quoted)
            {
                cells.add(cell.toString());
                cell.setLength(0);
            }
            else
            {
                cell.append(value);
            }
        }
        cells.add(cell.toString());
        return cells;
    }

    private static String csvCell(String value)
    {
        if (value.indexOf(',') < 0 && value.indexOf('"') < 0 && value.indexOf('\n') < 0)
        {
            return value;
        }
        return '"' + value.replace("\"", "\"\"") + '"';
    }

    private static boolean toBoolean(String value)
    {
        return switch (value.strip().toLowerCase(Locale.ROOT))
        {
            case "1", "true", "yes", "y" -> true;
            default -> false;
        };
    }

    public static void main(String[] arguments)
    {
        int exitCode = GeneratedWindowRuntime.runGeneratedAppMain(() ->
        {
            ContactsManager application = new ContactsManager();
            return GeneratedWindowRuntime.runGeneratedWindow(
                application,
                application.runtimeOptions(),
                arguments);
        });
        System.exit(exitCode);
    }

    private record Contact(
        String id,
        String firstName,
        String lastName,
        String email,
        String phone,
        String company,
        String city,
        String category,
        boolean active,
        String notes)
    {
        Contact withId(String value)
        {
            return new Contact(
                value,
                firstName,
                lastName,
                email,
                phone,
                company,
                city,
                category,
                active,
                notes);
        }
    }

    private static final class ContactForm extends ContactFormUI
    {
        private String contactId = "";
        private Consumer<Contact> onClosed;

        ContactForm(Contact contact)
        {
            if (contact == null)
            {
                return;
            }
            contactId = contact.id();
            first_name.setValue(contact.firstName());
            last_name.setValue(contact.lastName());
            email.setValue(contact.email());
            phone.setValue(contact.phone());
            company.setValue(contact.company());
            city.setValue(contact.city());
            selectCategory(contact.category());
            active.setChecked(contact.active());
            notes.setValue(contact.notes());
        }

        void setOnClosed(Consumer<Contact> value)
        {
            onClosed = value;
        }

        @Override
        protected void onSaveBtnClick()
        {
            if (first_name.value().isBlank()
                || last_name.value().isBlank()
                || email.value().isBlank())
            {
                message.setText("Required fields are missing.");
                return;
            }
            if (onClosed != null)
            {
                onClosed.accept(new Contact(
                    contactId,
                    first_name.value(),
                    last_name.value(),
                    email.value(),
                    phone.value(),
                    company.value(),
                    city.value(),
                    category.selectedText(),
                    active.checked(),
                    notes.value()));
            }
            closeWindow();
        }

        @Override
        protected void onCancelBtnClick()
        {
            if (onClosed != null)
            {
                onClosed.accept(null);
            }
            closeWindow();
        }

        private void selectCategory(String value)
        {
            int index = category.options().indexOf(value);
            category.setSelectedIndex(index >= 0 ? index : 0);
        }
    }
}
